// ts-to-zod/lib/core/jsDocTags.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.jsDocTagToZodProperties = exports.getJSDocTags = void 0;
const tslib_1 = require("tslib");
const typescript_1 = tslib_1.__importDefault(require("typescript"));
const tsutils_1 = require("tsutils");
const { factory: f } = typescript_1.default;
/**
 * List of formats that can be translated in zod functions.
 */
const formats = [
    "email",
    "uuid",
    // "uri" as const,
    "url",
    // "date" as const,
    // "date-time" as const,
];
/**
 * Typeguard to filter supported JSDoc tag key.
 *
 * @param tagName
 */
function isJSDocTagKey(tagName) {
    const keys = [
        "minimum",
        "maximum",
        "default",
        "minLength",
        "maxLength",
        "format",
        "pattern",
    ];
    return keys.includes(tagName);
}
/**
 * Typeguard to filter supported JSDoc format tag values.
 *
 * @param format
 */
function isSupportedFormat(format = "") {
    return formats.includes(format);
}
/**
 * Return parsed JSTags.
 *
 * @param nodeType
 * @param sourceFile
 * @returns Tags list
 */
function getJSDocTags(nodeType, sourceFile) {
    const jsDoc = tsutils_1.getJsDoc(nodeType, sourceFile);
    const jsDocTags = {};
    if (jsDoc.length) {
        jsDoc.forEach((doc) => {
            (doc.tags || []).forEach((tag) => {
                const tagName = tag.tagName.escapedText.toString();
                if (!isJSDocTagKey(tagName))
                    return;
                switch (tagName) {
                    case "minimum":
                    case "maximum":
                    case "minLength":
                    case "maxLength":
                        if (tag.comment && !Number.isNaN(parseInt(tag.comment))) {
                            jsDocTags[tagName] = parseInt(tag.comment);
                        }
                        break;
                    case "pattern":
                        if (tag.comment) {
                            jsDocTags[tagName] = tag.comment;
                        }
                        break;
                    case "format":
                        if (isSupportedFormat(tag.comment)) {
                            jsDocTags[tagName] = tag.comment;
                        }
                        break;
                    case "default":
                        if (tag.comment &&
                            !tag.comment.includes('"') &&
                            !Number.isNaN(parseInt(tag.comment))) {
                            // number
                            jsDocTags[tagName] = parseInt(tag.comment);
                        }
                        else if (tag.comment && ["false", "true"].includes(tag.comment)) {
                            // boolean
                            jsDocTags[tagName] = tag.comment === "true";
                        }
                        else if (tag.comment &&
                            tag.comment.startsWith('"') &&
                            tag.comment.endsWith('"')) {
                            // string with double quotes
                            jsDocTags[tagName] = tag.comment.slice(1, -1);
                        }
                        else if (tag.comment) {
                            // string without quotes
                            jsDocTags[tagName] = tag.comment;
                        }
                        break;
                }
            });
        });
    }
    return jsDocTags;
}
exports.getJSDocTags = getJSDocTags;
/**
 * Convert a set of jsDocTags to zod properties
 *
 * @param jsDocTags
 * @param isOptional
 * @param isPartial
 * @param isRequired
 */
function jsDocTagToZodProperties(jsDocTags, isOptional, isPartial, isRequired) {
    const zodProperties = [];
    if (jsDocTags.minimum !== undefined) {
        zodProperties.push({
            identifier: "min",
            expressions: [f.createNumericLiteral(jsDocTags.minimum)],
        });
    }
    if (jsDocTags.maximum !== undefined) {
        zodProperties.push({
            identifier: "max",
            expressions: [f.createNumericLiteral(jsDocTags.maximum)],
        });
    }
    if (jsDocTags.minLength !== undefined) {
        zodProperties.push({
            identifier: "min",
            expressions: [f.createNumericLiteral(jsDocTags.minLength)],
        });
    }
    if (jsDocTags.maxLength !== undefined) {
        zodProperties.push({
            identifier: "max",
            expressions: [f.createNumericLiteral(jsDocTags.maxLength)],
        });
    }
    if (jsDocTags.format) {
        zodProperties.push({
            identifier: jsDocTags.format,
        });
    }
    if (jsDocTags.pattern) {
        zodProperties.push({
            identifier: "regex",
            expressions: [f.createRegularExpressionLiteral(`/${jsDocTags.pattern}/`)],
        });
    }
    if (isOptional) {
        zodProperties.push({
            identifier: "optional",
        });
    }
    if (isPartial) {
        zodProperties.push({
            identifier: "partial",
        });
    }
    if (isRequired) {
        zodProperties.push({
            identifier: "required",
        });
    }
    if (jsDocTags.default !== undefined) {
        zodProperties.push({
            identifier: "default",
            expressions: jsDocTags.default === true
                ? [f.createTrue()]
                : jsDocTags.default === false
                    ? [f.createFalse()]
                    : typeof jsDocTags.default === "number"
                        ? [f.createNumericLiteral(jsDocTags.default)]
                        : [f.createStringLiteral(jsDocTags.default)],
        });
    }
    return zodProperties;
}
exports.jsDocTagToZodProperties = jsDocTagToZodProperties;
